flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java

```diff
@@ -490,8 +490,11 @@ private void insertIntoInternal(UnresolvedIdentifier unresolvedIdentifier, Table
 	private Optional<CatalogQueryOperation> scanInternal(UnresolvedIdentifier identifier) {
 		ObjectIdentifier tableIdentifier = catalogManager.qualifyIdentifier(identifier);
 
-		return catalogManager.getTable(tableIdentifier)
-			.map(t -> new CatalogQueryOperation(tableIdentifier, t.getTable().getSchema()));
+		return catalogManager.getTable(tableIdentifier).map(t -> {
+			CatalogTableSchemaResolver resolver = new CatalogTableSchemaResolver(parser);
```

Comment:
Shouldn't this resolution be part of the `TableSchema` itself? IMO when a watermarkSpec is applied to a `TableSchema` it should adjust the type of the rowtime attribute field. I think it's too late in the `TableEnvironment` to apply that transformation

Reply:
We can fix the `rowtime` type in `TableSchema` based on `WatermarkSpec`, but we can‘t fix the `proctime` type, unless we just use string compare(`proctime()`). As discussed in https://github.com/apache/flink/pull/12260, we should use `SqlExprToRexConverter` to get the correct type. For table api, there is the entry point to create a flink `Table`, and for sql, `CatalogSchemaTable` is the entry point to create a calcite `Table`.

Reply:
I had a look at #12260 and my feeling is even stronger that fix it everywhere but not where it actually should be fixed. IMO if we fixed it when the Table is created and have the proper types from the very beginning we would not need to patch it everywhere.

I have the feeling that the problem here is that the type we store in the catalog is wrong. I looked into it and my take is that the TimestampType#kind should also be stored in the catalog and this would fix most of the problems. Right now we are basically stripping that information when storing it in catalog. I am not 100% sure if we should store the Time attribute property, but I think it would fix imo hacks in this PR and #12260. I'd like to know what @twalthr thinks about it.

Reply:
We had a long discussion in FLINK-17189 and had a conclusion that the TimestampType#kind shoudn't be stored in catalog. IMO, there are several reasons: 

1) the metadata stored in catalog should simply store the information of DDL. In other words, we should only store the pure DDL into catalog. 
2) the result datatype of computed column is not trusted, they may changed across versions, including `PROCTIME()` and user-defined functions. There are some users reported that the result type of `PROCTIME()` should be `TIMESTAMP WITH LOCAL TIME ZONE`, not `TIMESTAMP WITHOUT TIME ZONE`.  Users may also change their UDF implementation. If we just use the result type stored in catalog, there of course will be an exception during code generation or runtime, because of type mismatch. 
3) I don't want to treat proctime attribute and rowtime attribute specially when storing them in catalog. They are not different than other computed columns and regular columns. 

Thus, we **have to** resolve the result type of computed columns again, when we get the `CatalogTable` from catalog. Actually, the initial `CatalogTable` get from catalog is a unresovled table (like UnresolvedExpression). That's why we need a `CatalogTableShemaResolver` to resolve `CatalogTableSchema`.

Reply:
I am 90% on the same page with the resolution. The 10 remaining percent still think that the stored type should be correct, but let's leave it out for now and assume I agree with the additional resolution. 

I still think we can place the resolution at a better location to reuse it between Table API & SQL. Would it work if we put it in the `CatalogManager#getTable`? That way it would be adjusted both for the SQL & Table API.

Reply:
In the end my reasoning is that this logic should be pushed as close to the catalog as possible. If we cannot push it all the way to the creation of the table, maybe at least we can push it into the single place when we look up that table.

Reply:
Putting it in `CatalogManager#getTable` sounds good to me. We thought to have a single util to resolve the schema. But I think you have a better idea to resolve the schema before CatalogTable is exposed out of CatalogManager, this can avoid to forget resove the shema.

Reply:
Thanks for the suggestion. I'm also agree we move the resolution logic into `CatalogManager#getTable` as a unified place for Table API and SQL.